server/src/main/java/org/candlepin/async/StatusUpdatingJob.java

```diff
@@ -0,0 +1,74 @@
+/**
+ * Copyright (c) 2009 - 2019 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.async;
+
+import org.candlepin.model.AsyncJobStatus;
+import org.candlepin.model.AsyncJobStatusCurator;
+import org.candlepin.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Serializable;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Handles updates of job's {@link AsyncJobStatus}. Sets the state to
+ * JobState.RUNNING at the beginning of its execution. At the end updates to
+ * either COMPLETE or FAILED and sets the job result.
+ */
+public class StatusUpdatingJob implements AsyncJob {
+
+    public static final Logger log = LoggerFactory.getLogger(StatusUpdatingJob.class);
+
+    private final AsyncJob job;
+    private final Serializable jobId;
+    private final AsyncJobStatusCurator curator;
+
+    public StatusUpdatingJob(final AsyncJob job, final Serializable jobId,
+        final AsyncJobStatusCurator curator) {
+        this.job = requireNonNull(job);
+        this.jobId = requireNonNull(jobId);
+        this.curator = requireNonNull(curator);
+    }
+
+    @Override
+    public Object execute(final JobExecutionContext jdata) throws JobExecutionException {
+        setRunning();
+        try {
+            final Object result = job.execute(jdata);
+            updateStatus(AsyncJobStatus.JobState.COMPLETED, result);
+            return result;
+        }
+        catch (JobExecutionException e) {
```

Comment:
Is there any exception where the job should not be recorded as Failed?

Reply:
From looking at the old job implementation and KingpinJob it seems there is no such case.

Reply:
Let's have this catch any exception to mark the job as failed.

Reply:
Done.